### ISSUE TITLE
error in backend: 32 bit reloc applied to a field with a different size

### ISSUE THREAD

**Issue by practicalswift** [GJDuck/EffectiveSan#2]
When trying out EffectiveSan on the Bitcoin Core project I get the error message “error in backend: 32 bit reloc applied to a field with a different size”. What am I doing wrong? :-)

```
$ git clone https://github.com/bitcoin/bitcoin
$ cd bitcoin
$ ./autogen.sh
$ CC=/…/effectivesan-0.1-alpha/bin/clang CXX=/…/effectivesan-0.1-alpha/bin/clang++ \
    ./configure --with-incompatible-bdb --with-sanitizers=effective --disable-asm
$ make
…
fatal error: error in backend: 32 bit reloc applied to a field with a different size
Makefile:8532: recipe for target 'crypto/libbitcoinconsensus_la-hmac_sha512.lo' failed
make[2]: *** [crypto/libbitcoinconsensus_la-hmac_sha512.lo] Error 1
```

**Comment by GJDuck** [comment]
Normally these kinds of errors are caused by inline asm or linking against modules written in assembly.  However, I see that you have included the `--disable-asm` config option, so I am not sure.  It could be a new bug so I'll need to investigate.

In the meantime, disabling global variable instrumentation might work.  This can be achieved by passing the following options to the compiler:

    -mllvm -effective-no-globals -mcmodel=small

**Comment by GJDuck** [comment]
OK I managed to build it by disabling global instrumentation as suggested above:

    CC=~/effectivesan-0.1-alpha/bin/clang CXX=~/effectivesan-0.1-alpha/bin/clang++  CXXFLAGS="-mllvm -effective-no-globals -mcmodel=small -O2" CFLAGS="-mllvm -effective-no-globals -mcmodel=small -O2"  ./configure --with-incompatible-bdb --with-sanitizers=effective --disable-asm

Maybe it is still using assembly despite the `--disable-asm` config option?

Running `bitcoind` produced a few errors but mainly relating to `std::_Rb_tree_node` (see FAQ in README).

**Comment by practicalswift** [comment]
@GJDuck Thanks! The project now compiles under EffectiveSan. Cool!

I'm now encountering the following error:

```
$ src/bitcoind
_|                                      _|_|_|_|            _|
_|          _|_|    _|      _|      _|  _|        _|_|_|  _|_|_|_|
_|        _|    _|  _|      _|      _|  _|_|_|  _|    _|    _|
_|        _|    _|    _|  _|  _|  _|    _|      _|    _|    _|
_|_|_|_|    _|_|        _|      _|      _|        _|_|_|      _|_|

LOWFAT ERROR: incompatible architecture (no BMI/BMI2 support)
Aborted
```

CPU:

```
$ grep "model name" /proc/cpuinfo  | head -1
model name      : Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz
```

Is this expected? :-)

**Comment by practicalswift** [comment]
When trying to build from source I encounter the following issue. What am I doing wrong? :-)

```
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
$ git clone https://github.com/GJDuck/EffectiveSan
$ cd EffectiveSan/
$ ./build.sh release
…
./build.sh: checking the CPU...
./build.sh: warning: CPU does not support BMI

./build.sh: warning: CPU does not support BMI2

./build.sh: warning: CPU does not support LZCNT
./build.sh: building the LowFat config builder...
./build.sh: building the LowFat config...
./build.sh: building the LowFat config check...
./build.sh: checking the LowFat config...
./build.sh: copying the LowFat config files...
./build.sh: creating LLVM build directory (build)...
./build.sh: installing the LowFat ld script file...
./build.sh: will now build LLVM...
-- The C compiler identification is Clang 4.0.1
-- The CXX compiler identification is Clang 4.0.1
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang-4.0
-- Check for working C compiler: /usr/bin/clang-4.0
-- Check for working C compiler: /usr/bin/clang-4.0 -- works
…
-- Clang version: 4.0.1
-- Performing Test CXX_SUPPORTS_NO_NESTED_ANON_TYPES_FLAG
-- Performing Test CXX_SUPPORTS_NO_NESTED_ANON_TYPES_FLAG - Success
-- Configuring done
CMake Error at projects/compiler-rt/cmake/Modules/AddCompilerRT.cmake:189 (add_library):
  Cannot find source file:

    lowfat.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  projects/compiler-rt/lib/effective/CMakeLists.txt:14 (add_compiler_rt_runtime)


CMake Error: Cannot determine link language for target "clang_rt.effective-x86_64".
CMake Error: CMake can not determine linker language for target: clang_rt.effective-x86_64
-- Generating done
-- Build files have been written to: /…/EffectiveSan/build
$ cd build
$ make
/…/EffectiveSan/llvm-4.0.1.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_stoptheworld_linux_libcdep.cc:278:22: error: variable has incomplete type 'struct sigaltstack'
  struct sigaltstack handler_stack;
                     ^
/…/EffectiveSan/llvm-4.0.1.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_linux.h:24:8: note: forward declaration of 'sigaltstack'
struct sigaltstack;
       ^
1 error generated.
projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/build.make:254: recipe for target 'projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/sanitizer_stoptheworld_linux_libcdep.cc.o' failed
make[2]: *** [projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/sanitizer_stoptheworld_linux_libcdep.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by GJDuck** [comment]
Somehow the `build.sh` script was broken.  It [should now be fixed](https://github.com/GJDuck/EffectiveSan/commit/6e58ac804645c7c5c6af21cf2cc4be87b4ae0127).

**Comment by practicalswift** [comment]
Thanks for the quick turnaround time! I tried again and got:

```
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
$ git clone https://github.com/GJDuck/EffectiveSan
$ cd EffectiveSan/
$ ./build.sh release
…
/…/EffectiveSan/llvm-4.0.1.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_stoptheworld_linux_libcdep.cc:278:22: error: variable has incomplete type 'struct sigaltstack'
  struct sigaltstack handler_stack;

                     ^
/…/EffectiveSan/llvm-4.0.1.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_linux.h:24:8: note: forward declaration of 'sigaltstack'
struct sigaltstack;
       ^
1 error generated.
projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/build.make:254: recipe for target 'projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/sanitizer_stoptheworld_linux_libcdep.cc.o' failed
make[2]: *** [projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/sanitizer_stoptheworld_linux_libcdep.cc.o] Error 1
CMakeFiles/Makefile2:6817: recipe for target 'projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/all' failed
make[1]: *** [projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonLibc.x86_64.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
…
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by GJDuck** [comment]
It's building fine for me, but I'm still using 16.04.  I can only guess the system headers have changed and the definition of `struct sigaltstack` has moved?

Perhaps try some nasty fix, e.g., copy the definition into `sanitizer_stoptheworld_linux_libcdep.cc` before the error:
 
    struct sigaltstack { void *ss_sp; int ss_flags; size_t ss_size; };

Then:

    rm -rf build
    ./build.sh release

Otherwise I'll try to set up a 18.04 VM and test.

**Comment by GJDuck** [comment]
The `sigaltstack` problem should hopefully be fixed by 65dde63.

I had the same problem with LowFat & had fixed it a while ago.  I evidently forgot to apply the same patch to the EffectiveSan repo.

**Comment by practicalswift** [comment]
Ah, got it! I tried lastest `master` and got:

```
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
$ git clone https://github.com/GJDuck/EffectiveSan
$ cd EffectiveSan/
$ ./build.sh release
…
[ 70%] Building CXX object lib/Linker/CMakeFiles/LLVMLinker.dir/LinkModules.cpp.o
In file included from /root/EffectiveSan/llvm-4.0.1.src/lib/Transforms/Instrumentation/AddressSanitizer.cpp:16:
In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/ArrayRef.h:13:
In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/Hashing.h:49:
In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/Support/Host.h:17:
In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/StringMap.h:17:
In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/StringRef.h:13:
In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/STLExtras.h:24:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/functional:54:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/tuple:39:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/array:39:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/stdexcept:39:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/string:52:
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:113:15: error: unknown type name 'basic_string_view'
      typedef basic_string_view<_CharT, _Traits> __sv_type;
              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:113:32: error: expected member name or ';' after declaration specifiers
      typedef basic_string_view<_CharT, _Traits> __sv_type;
      ~~~~~~~~~~~~~~~~~~~~~~~~~^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:117:38: error: use of undeclared identifier '__sv_type'
          __and_<is_convertible<const _Tp&, __sv_type>,
                                            ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:117:48: error: expected a type
          __and_<is_convertible<const _Tp&, __sv_type>,
                                                      ^
…
```

**Comment by GJDuck** [comment]
This one I am not too sure about, it seems to be a problem with the c++ standard headers.  I might have to set up a 18.04 VM and test for myself.

As an educated guess, you could try adding the following line to the top of `EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/STLExtras.h`:

    #include <string_view>

**Comment by GJDuck** [comment]
After examining the code base, rebuilding might not work anyway.  I think I'll just remove the dependency on modern CPU features and make a new release.

**Comment by GJDuck** [comment]
I've released a new version that should be compatible with older CPUs:

https://github.com/GJDuck/EffectiveSan/releases/tag/v0.1.1-alpha

I've verified that the test suite works on a vintage Intel i5-2500K (cira 2011), although I have not tried spec or firefox yet.

**Comment by practicalswift** [comment]
@GJDuck Thanks! Now working as expected.

**Comment by mansouriahmad** [comment]
> Ah, got it! I tried lastest `master` and got:
> 
> ```
> $ cat /etc/lsb-release
> DISTRIB_ID=Ubuntu
> DISTRIB_RELEASE=18.04
> DISTRIB_CODENAME=bionic
> DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
> $ git clone https://github.com/GJDuck/EffectiveSan
> $ cd EffectiveSan/
> $ ./build.sh release
> …
> [ 70%] Building CXX object lib/Linker/CMakeFiles/LLVMLinker.dir/LinkModules.cpp.o
> In file included from /root/EffectiveSan/llvm-4.0.1.src/lib/Transforms/Instrumentation/AddressSanitizer.cpp:16:
> In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/ArrayRef.h:13:
> In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/Hashing.h:49:
> In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/Support/Host.h:17:
> In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/StringMap.h:17:
> In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/StringRef.h:13:
> In file included from /root/EffectiveSan/llvm-4.0.1.src/include/llvm/ADT/STLExtras.h:24:
> In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/functional:54:
> In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/tuple:39:
> In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/array:39:
> In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/stdexcept:39:
> In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/string:52:
> /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:113:15: error: unknown type name 'basic_string_view'
>       typedef basic_string_view<_CharT, _Traits> __sv_type;
>               ^
> /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:113:32: error: expected member name or ';' after declaration specifiers
>       typedef basic_string_view<_CharT, _Traits> __sv_type;
>       ~~~~~~~~~~~~~~~~~~~~~~~~~^
> /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:117:38: error: use of undeclared identifier '__sv_type'
>           __and_<is_convertible<const _Tp&, __sv_type>,
>                                             ^
> /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:117:48: error: expected a type
>           __and_<is_convertible<const _Tp&, __sv_type>,
>                                                       ^
> …
> ```

Hi Gregory ,

I am getting the same error compiling latest version of EffectiveSan on ubuntu. Clang-4.0 is installed too.
